Netatmo migration tool to migrate old and new devices

LoukKamp opened the pull request:
First beta release for testing and discussion.

Merged with Netatmo rework by @JurgenLB.

Note: This version is only tested with the Netatmo Weather Station. Other devices still need testing or even adjusting when device units, types, or subtypes are changed.

LoukKamp committed:
First update with Netatmod device migrator, just testing and selecting: No updates done

LoukKamp committed:
Merge branch 'domoticz:development' into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Merge branch 'domoticz:development' into Netatmo-rework-by-JurgenJB

LoukKamp committed:
2nd update incl.Migragtion flag on Netatmo hardware page and option to force all data copied

LoukKamp committed:
Merge branch 'Netatmo-rework-by-JurgenJB' of https://github.com/LoukKamp/myOwnDomoticz into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Merge branch 'domoticz:development' into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Merge branch 'domoticz:development' into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Final beta to be tested with other devices than the weather station

LoukKamp committed:
Merge branch 'Netatmo-rework-by-JurgenJB' of https://github.com/LoukKamp/myOwnDomoticz into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Merge branch 'Netatmo-rework-by-JurgenJB' of https://github.com/LoukKamp/myOwnDomoticz into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Remove inrellevant differences from Netatmo.cpp

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -488,14 +526,146 @@ uint64_t CNetatmo::convert_mac(std::string mac)
 }
 
 
+uint64_t getValueFromString(const std::string& value)
```
Instead of this function please use **std::stoll**

LoukKamp replied:
OK

LoukKamp replied:
Done

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -488,14 +526,146 @@ uint64_t CNetatmo::convert_mac(std::string mac)
+	return result;
+}
+
+std::string getStringFromValue(const uint64_t value)
```
Instead of this function, please use **std::to_string**

LoukKamp replied:
OK

LoukKamp replied:
Done

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -488,14 +526,146 @@ uint64_t CNetatmo::convert_mac(std::string mac)
+/// updating the old device to the new devices and finaly remove the new device.
+/// Note: the old device will become the new device.
+/// </summary>
+bool CNetatmo::MergeDevices (const uint64_t ipOldDeviceId, const uint64_t ipNewDeviceId)
```
Please remove the space after the function name

LoukKamp replied:
OK

LoukKamp replied:
Done

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -488,14 +526,146 @@ uint64_t CNetatmo::convert_mac(std::string mac)
+
+	Log(LOG_STATUS, "MergeDevices: Merge old device %s with new device %s", sOldDeviceId.c_str(), sNewDeviceId.c_str());
+
+	auto result = m_sql.safe_query
```
No need to apply the end semicolon, and it could be a simple one line

LoukKamp replied:
OK

LoukKamp replied:
On the second hand, I don't see what you mean here: There is only one semicolon in the (single) line with the log statement, and it is required by the C++ standard...

gizmocuz replied:
It's in the sql query

LoukKamp replied:
I removed the query, so problem solved. 

(I also removed the semicolons from the other SQL statements)

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -488,14 +526,146 @@ uint64_t CNetatmo::convert_mac(std::string mac)
+		sOldDeviceId.c_str(),
+		sNewDeviceId.c_str()
+	);
+	if (result.empty())
```
Here you check if the result is empty, and the next check you check if the result contains at least two values.
This function is only called on certain conditions below, and only by this class, so I think the first check can be removed, even the second check i guess... 

Also please keep the variable names case sensitive as how there are created. Id --> ID

LoukKamp replied:
Investigating....

LoukKamp replied:
These lines were meant as general check on all kinds of (external caused) errors (like e.g. a database lock) and a sanity check. I was learned to test always on the unexpected...

I'll remove them, because there is no need for this query anymore.

LoukKamp replied:
Done (removed)

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -488,14 +526,146 @@ uint64_t CNetatmo::convert_mac(std::string mac)
+{
+	auto logResult = m_sql.safe_query
+	(
+		"SELECT COUNT(*) FROM %q as a  WHERE a.deviceRowId='%q' AND  EXISTS (SELECT date FROM %q as b WHERE b.DeviceRowID = '%q' AND b.date == a.date);",
```
Please keep the database column names case sensitive, how they are defined. like DeviceRowID, Date etc.

gizmocuz replied:
Also for the remaining code please

LoukKamp replied:
OK

LoukKamp replied:
Done

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -504,8 +674,167 @@ uint64_t CNetatmo::UpdateValueInt(int HardwareID, const char* ID, unsigned char
         }
         m_mainworker.sOnDeviceReceived(m_HwdID, DeviceRowIdx, devname, nullptr);
-        m_notifications.CheckAndHandleNotification(DeviceRowIdx, m_HwdID, ID, devname, unit, devType, subType, nValue, sValue);
+        m_notifications.CheckAndHandleNotification(DeviceRowIdx, m_HwdID, sDeviceID.c_str(), devname, unit, devType, subType, nValue, sValue);
```
no need for the c_str() here

LoukKamp replied:
This is probably Jurgen's part of the code, but I'll take a look.

LoukKamp replied:
In the end, it was me: I've changed this because of the upper case conversion needed to compare the device IDs. Indeed, the c_str() function is not needed, as the parameter for the called function is also std::string.

But now I realize that this converted uppercase HEX value is also stored in the database, and this was not the intention of the author (@JurgenLB). So when he wants this to be corrected, I need to see if this has any consequences on the matching process: Original device IDs were in the format "04X" (4 uppercase HEX digits).

LoukKamp replied:
I moved the uppercase translation, so the updates are done without it, but found that @JurgenLB already changed this in his part of the code. So, in fact, this change has no effect on the results: I'm happy with it.

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -504,8 +674,167 @@ uint64_t CNetatmo::UpdateValueInt(int HardwareID, const char* ID, unsigned char
+		std::vector<std::vector<std::string> > result;
+		std::string searchString = sDeviceID.substr(sDeviceID.length()-2,2);
+
+		result = m_sql.safe_query(
```
Please keep the MySQL functions case sensitive, like most of them are in CAPITAL letters, see the documentation

https://www.w3schools.com/sql/func_mysql_substr.asp

Also check this throughout this PR

LoukKamp replied:
OK (I suppose you mean SUBSTR)

gizmocuz replied:
correct and also some date/time functions, most things are case sensitive, like SELECT instead of Select for instance (not that you did this... )

LoukKamp replied:
The example above is not a SQL syntax, but C++, so it needs to be in lower case in this case. However, I modified all other cases where it was part of a SQL statement.

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -504,8 +674,167 @@ uint64_t CNetatmo::UpdateValueInt(int HardwareID, const char* ID, unsigned char
+						MergeDevices(oldID, DeviceRowIdx);
+						break; 	//One device per cycle
+					}
+					else if (oldUnit == unit && oldDevType == 82 && oldDevType == devType && oldSubType == 5 && subType == 113)
```
As these if/then/else lines can become quite large, I use another coding formatting throughout the Domoticz code like
```

    else if (
        (oldUnit == unit)
        && (oldDevType == 82)
        && (oldDevType == devType)
        && (oldSubType == 5)
        && (subType == 113)
        )
    {
        ....
    }

```

I think this also looks a lot cleaner

And instead of these numbers, please use the real variable/defines, like defined in **hardwaretypes.h** or **RFXtrx.h**

82 -> pTypeTEMP_HUM

Numbers 'can' change, and nobody knows whats happening by looking at just numbers

LoukKamp replied:
OK, but I looked in hardwaretypes.h, but didn't find matching names for the numbers. Maybe in RFXtrx.h; I'll take a look.

LoukKamp replied:
Done (both requests)

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -1872,10 +2202,19 @@ bool CNetatmo::ParseDashboard(const Json::Value& root, const int DevIdx, const i
 		// check if Netatmo data was updated in the past NETAMO_POLL_INTERVALL (+1 min for sync time lags)... if not means sensors failed to send to cloud
 		int Interval = NETAMO_POLL_INTERVALL + 60;
 		//Debug(DEBUG_HARDWARE, "Module [%s] Interval = %d", name.c_str(), Interval);
+		Log(LOG_STATUS, "Module [%s] Interval = %d", name.c_str(), Interval);
```
I think this should be a DEBUG_HARDWARE ?

LoukKamp replied:
Sorry, but I (miss-)use LOG_STATUS for logging/debugging, because when I enable debugging, I get a lot of debug messages from other devices that will distract my attention of the important message to understand what is happening.

I will remove them later (when I get some results from other beta testers with devices I don't own).

gizmocuz replied:
if you want you can also add

#ifdef _DEBUG
... do something
#endif

Then this will only be displayed when you build in debug mode (which you do while developing)

LoukKamp replied:
That would not help, because after merging to the beta release, it will be built without the _DEBUG value defined. I will need this information to check the logs from (BETA-) users that experience issues or have devices I don't own. 

By the way, I don't have a proper debug environment (with a decent symbolic) setup yet, because the first trials to do so failed, and I didn't have the time to figure that out. So no useful need for debug mode so far... Sorry!

gizmocuz replied:
Maybe try to install/use visual studio? Community version is free and great!

LoukKamp replied:
Thanks for the tip: I used this for several years to development in C# on a Windows system. But can this be used to debug on a Raspberry Pi, running Linux? In need to figure this out, but too late for this project, I guess.

gizmocuz commented on hardware/Netatmo.h:
```diff
@@ -166,4 +168,7 @@ class CNetatmo : public CDomoticzHardwareBase
 	uint64_t UpdateValueInt(int HardwareID, const char* ID, unsigned char unit, unsigned char devType, unsigned char subType, unsigned char signallevel, unsigned char batterylevel, int nValue, const char* sValue, std::string& devname, bool bUseOnOffAction, const std::string& user);
 	bool ParseDashboard(const Json::Value &root, int DevIdx, int ID, std::string& name, const std::string &ModuleType, int battery_percent, int rf_status, std::string& Hardware_ID, std::string& home_id);
+
+	bool MergeDevices (const uint64_t ipOldDeviceId, const uint64_t ipNewDeviceId);
```
Please remove the space behind the function name

LoukKamp replied:
OK (sorry, was not my intention ;)).

LoukKamp replied:
Done

gizmocuz commented on www/app/hardware/Hardware.html:
```diff
@@ -1485,6 +1485,10 @@ <h2><span data-i18n="Hardware">Hardware</span>:</h2>
 				</td>
 			</tr>
+			<tr>
+				<td align="right" style="width:110px"><span data-i18n="Migration Tool"></span>:</label></td>
```
I guess the migration is a onetime only thing to do?
If yes, this should not be here.
Maybe you could use a Preference version value (NetatmoHWVersion) and if this does not exists, or is a specific value, call the migration at startup of the Netatmo hardware?

gizmocuz replied:
This way also no changes in the web part needs to be made

LoukKamp replied:
Not really: It needs to run several times, depending on the moment that the "Early Adaptor" started using the new software (there were several changes over time in the device ID (incl. case changes), unit numbers, types, and subtypes during the development process). I planned to do this in one run, but realized that the "New" device (that is triggering the migration for this device) was changing during the process, and also, I needed to work backwards in time. So I needed to split up. Users that will step in now, probably will need only one run, but users that stepped in after the first update, probably need to run it 3 or even 4 times (as in my case).

I also think that I should give the user an option to use the tool or not (the default setting of the flag is OFF).

There is no changed hardware or firmware version to test on; All changes are caused by the new interface within Domoticz (in hardware/Netatmo.cpp).

The only way to test if a migration is useful, is the very complex SQL statement that matches old and new devices, based on their Device ID (or a part of it). That is why I wanted to introduce the flag, so it can be switched off after the conversion to avoid unnecessary database activity.

I also think that the whole tool, including the flag on the website, can be removed after one or two master releases.

gizmocuz replied:
Yes, that's why I mentioned the idea of using a version number in the preferences so you can migrate on startup of the hardware

LoukKamp replied:
But, in that case, I need to add an option to the preferences' webpage to enable the migration. While in BETA, I don't want to run this tool without permission and knowledge of the user.

gizmocuz replied:
No, there are functions like

`CSQLHelper::GetPreferencesVar`

```
m_netamo_hw_version = 1;
if (!m_sql.GetPreferencesVar("NetamoHWVersion", m_netamo_hw_version)) m_netamo_hw_version = 0;

```

LoukKamp replied:
That still leave me with two questions: How should the user enable the migration tool, and what if the user has more versions of the device running (e.g. one for production and one for testing like me, or one for the weather station and one for the alarm system)?

Can we please agree that I use this flag on the hardware page for beta testing, and that I replace it with your suggestion (or something like that) before the next master release?

gizmocuz replied:
The user should 'not' be bothered with the migration tool. You as developer knows this.
Version 2 (this new version), we need to convert some sensors.
Later some more changes, version 3 (#define at the top of the class), you need to convert more

LoukKamp replied:
OK, I rest my case: see the last commit (removed migration flag)

gizmocuz reviewed (changes requested):
Sorry, quite some remarks here

LoukKamp committed:
Changes requested in review (Hardware/Netatmo.cpp and .h)

LoukKamp committed:
Second update after review comments (Remove uppercase befor update, and remove redundant semicolons in SQL statements

LoukKamp committed:
User settable migration flag removed and replace by automated preference
When this preference (NetamoProtVersion_ID_<hardware_ID> is not set,
version 0 is assumed and migration will be enabled until all devices
are migrated. After migration, NetamoProtVersion_ID_<hardware_ID> will
be updated to the next version (1 currently).
When there is an additional change in one or more device IDs, this
protocol verstion need to be adjusted in the source (Netatmo.cpp)

LoukKamp committed:
Spilt Migration tool from UpdateValueInt() into its own function and move init of migrationFlag from constructor to worker thread
The migration flag does not seem to be initialized properly from the class's constructor

LoukKamp committed:
Merge branch 'domoticz:development' into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Preview changes for migration tool (to diff)

LoukKamp committed:
Merge branch 'domoticz:development' into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Added prot version 2 (reversed device ID's and unit numbers (back to originals)

LoukKamp committed:
Merge branch 'Netatmo-rework-by-JurgenJB' of https://github.com/LoukKamp/myOwnDomoticz into Netatmo-rework-by-JurgenJB

LoukKamp committed:
Correct unintended change / merge

LoukKamp committed:
Undo percentage sensor changes for RF levels and remove commented remarks

LoukKamp committed:
Bugfix: CostomSensor should have custom type: sTypeCustom

LoukKamp committed:
Prepare merge with JurgenJB

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -53,6 +53,23 @@ std::string ReadFile(std::string filename)
 }
 #endif
 
+std::string prettifyJson(const Json::Value value)
```
Could you uncomment this function as well as it is not used?
Or add a #define at the top to enable/disable debugging, or make it a member function of the hardware class please

LoukKamp replied:
Added #ifdef _DEBUG as it is only used for debugging purposes

gizmocuz commented on hardware/Netatmo.cpp:
```diff
@@ -504,12 +576,445 @@ uint64_t CNetatmo::UpdateValueInt(int HardwareID, const char* ID, unsigned char
+/// <summary>
+/// Merge logs in all log tables
+/// </summary>
+bool CNetatmo::MergeDeviceLogs (const std::string& spTableName, const std::string& spOldDeviceId, const std::string& spNewDeviceId)
```
Could you remove the space behind the function name, or in visual studio press control-k followed by normal 'd'

LoukKamp replied:
Done

gizmocuz reviewed (approved):
Looking great!

LoukKamp committed:
Implement review results

LoukKamp committed:
Another merge with JurgenJB's code (comments only)

LoukKamp committed:
Merge branch 'Netatmo-rework-by-JurgenJB' of https://github.com/LoukKamp/myOwnDomoticz into Netatmo-rework-by-JurgenJB
Merge conflicts resolved

LoukKamp committed:
Re-add lost line (comment only)

LoukKamp committed:
Replace unit numbers according the type of temp sensor (some are fixed to 0 or 1, or extracted from the device ID (the last two digits when presented in HEX format)

LoukKamp committed:
Processed review comments from Jurgen

LoukKamp committed:
Processed review comments by Juregen (2/2)
